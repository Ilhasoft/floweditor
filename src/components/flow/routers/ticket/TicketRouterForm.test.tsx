import { mock } from 'testUtils';
import { createOpenTicketNode, getRouterFormProps } from 'testUtils/assetCreators';
import { Types } from 'config/interfaces';
import { RenderNode } from 'store/flowContext';
import TicketRouterForm from './TicketRouterForm';
import * as utils from 'utils';
import * as React from 'react';
import {
  render,
  fireEvent,
  fireUnnnicInputChangeText,
  fireUnnnicSelect,
  fireUnnnicAutocompleteSelect,
  act
} from 'test/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

// eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
const ticketForm = getRouterFormProps({
  node: createOpenTicketNode('Need help', 'Where are my cookies'),
  ui: { type: Types.split_by_ticket }
} as RenderNode);

describe(TicketRouterForm.name, () => {
  describe('render', () => {
    it('should render', async () => {
      let rendered: any;
      await act(async () => {
        rendered = render(<TicketRouterForm {...ticketForm} />);
      });
      expect(rendered.baseElement).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should save changes', async () => {
      let rendered: any;
      await act(async () => {
        rendered = render(<TicketRouterForm {...ticketForm} />);
      });

      expect(rendered.baseElement).toMatchSnapshot();

      const okButton = rendered.getByText('Ok');
      const resultName = rendered.getByTestId('Save as result');

      // our ticketer, body and result name are required
      fireUnnnicInputChangeText(resultName, '');
      fireEvent.click(okButton);
      expect(ticketForm.updateRouter).not.toBeCalled();

      fireUnnnicSelect(
        rendered.getByTestId('temba_select_assignee'),
        {
          email: 'agent.user@gmail.com',
          first_name: 'Agent',
          last_name: 'User',
          role: 'agent',
          created_on: '2021-06-10T21:44:30.971221Z'
        },
        'email'
      );

      // we need a topic
      fireUnnnicAutocompleteSelect(
        rendered.getByTestId('temba_select_topic'),
        {
          name: 'General',
          uuid: '6f38eba0-d673-4a35-82df-21bae2b6d466'
        },
        'name'
      );

      fireUnnnicInputChangeText(resultName, 'My Ticket Result');

      fireEvent.click(okButton);
      expect(ticketForm.updateRouter).toBeCalled();
      expect(ticketForm.updateRouter).toMatchCallSnapshot();
    });
  });
});
