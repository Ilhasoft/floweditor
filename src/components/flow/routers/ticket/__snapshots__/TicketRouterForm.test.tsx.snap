// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketRouterForm render should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header open_ticket"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Open Ticket
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"open_ticket\\",\\"name\\":\\"Open Ticket\\",\\"description\\":\\"Open a ticket with a human agent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"split_by_ticket\\"],\\"visibility\\":[\\"message\\",\\"voice\\"],\\"filter\\":\\"ticketer\\"}]"
              />
            </div>
          </div>
        </div>
        <p>
          <span>
            Open ticket via... 
          </span>
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_ticketer"
              errors="[]"
              name="Ticketer"
              namekey="name"
              placeholder="Select the ticketing service to use"
              valuekey="uuid"
              values="[{\\"name\\":\\"Email (bob@acme.com)\\",\\"uuid\\":\\"1165a73a-2ee0-4891-895e-768645194862\\"}]"
            />
          </div>
        </div>
        <div
          class="subject"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="Subject"
                placeholder="Enter a subject"
                type="text"
                value="Need help"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="body"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <textarea
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="Body"
                placeholder="Enter a body"
              >
                Where are my cookies
              </textarea>
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="64"
                name="Result Name"
                type="text"
                value="Result"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TicketRouterForm updates should save changes 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header open_ticket"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Open Ticket
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"open_ticket\\",\\"name\\":\\"Open Ticket\\",\\"description\\":\\"Open a ticket with a human agent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"split_by_ticket\\"],\\"visibility\\":[\\"message\\",\\"voice\\"],\\"filter\\":\\"ticketer\\"}]"
              />
            </div>
          </div>
        </div>
        <p>
          <span>
            Open ticket via... 
          </span>
        </p>
        <div
          class="ele"
        >
          <div
            class="normal "
          >
            <temba-select
              cachekey="0"
              data-testid="temba_select_ticketer"
              errors="[]"
              name="Ticketer"
              namekey="name"
              placeholder="Select the ticketing service to use"
              valuekey="uuid"
              values="[{\\"name\\":\\"Email (bob@acme.com)\\",\\"uuid\\":\\"1165a73a-2ee0-4891-895e-768645194862\\"}]"
            />
          </div>
        </div>
        <div
          class="subject"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="Subject"
                placeholder="Enter a subject"
                type="text"
                value="Need help"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="body"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <textarea
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="Body"
                placeholder="Enter a body"
              >
                Where are my cookies
              </textarea>
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="64"
                name="Result Name"
                type="text"
                value="Result"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TicketRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "body": "My Ticket Result",
          "result_name": "Result",
          "subject": "Need help",
          "ticketer": Object {
            "name": "Email (bob@acme.com)",
            "uuid": "1165a73a-2ee0-4891-895e-768645194862",
          },
          "type": "open_ticket",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_category",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.result",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_ticket",
    },
  },
]
`;
